File: frontend_svelte/src/lib/service/ssa.ts
import { apiUrl } from '$lib/config/api';
import type { GetNameStatsQuery } from '$lib/model/ssa/request';
import type { GetNameStatsResponse } from '$lib/model/ssa/response';
import type { TabularDataCollection, TabularDataRow } from '$lib/model/tabular';
import axios, { type AxiosRequestConfig } from 'axios';

const ssaApiUrl = `${apiUrl}/ssa`;

export async function getNameStats(query: GetNameStatsQuery): Promise<TabularDataCollection> {
	let response: GetNameStatsResponse;

	try {
		response = (await axios.post(`${ssaApiUrl}/name-stats`, query)).data;
	} catch (err) {
		console.error('GetNameStats request failed', err);
		throw err;
	}

	const name_sex_set: Set<string> = new Set();
	const year_row_map: { [year: number]: TabularDataRow } = {};

	response.forEach((item) => {
		const name_sex = `${item.name} (${item.sex})`;
		const row = (year_row_map[item.year] ??= { year: item.year });

		name_sex_set.add(name_sex);
		row[name_sex] = item.count;
	});

	const rows = Object.values(year_row_map);
	rows.sort((a, b) => a.year - b.year);

	const tabular_data: TabularDataCollection = {
		properties: name_sex_set,
		rows
	};

	return tabular_data;
}

File: backend_axum/src/controller/ssa.rs
use axum::{Extension, Json, Router, extract::State, routing::post};

use crate::{
    database,
    error::AppError,
    model::{
        auth::UserAuth,
        ssa::{request::GetNameStatsQuery, response::GetNameStatsResponse},
    },
    state::AppState,
};

pub fn router() -> Router<AppState> {
    Router::new().route("/name-stats", post(get_name_stats))
}

#[utoipa::path(
    get,
    path = "/ssa/name-stats",
    responses(
        (status = 200, description = "Yearly count of SSA registered names", body = GetNameStatsResponse),
        (status = BAD_REQUEST),
        (status = UNAUTHORIZED),
        (status = UNSUPPORTED_MEDIA_TYPE),
        (status = UNPROCESSABLE_ENTITY),
        (status = INTERNAL_SERVER_ERROR)
    ),
    request_body(content = GetNameStatsQuery, content_type = "application/json")
)]
#[tracing::instrument(skip(state))]
async fn get_name_stats(
    Extension(user_auth): Extension<UserAuth>,
    State(state): State<AppState>,
    Json(query): Json<GetNameStatsQuery>,
) -> Result<Json<GetNameStatsResponse>, AppError> {
    if !user_auth.permissions.iter().any(|item| item.as_str() == "read:ssa_stats") {
        return Err(AppError::Forbidden(
            "read:ssa_stats permission required".into(),
        ));
    }
    
    if query.start_year > query.end_year {
        return Err(AppError::UnprocessableEntity(
            "Start year must not be after end year".into(),
        ));
    }

    let database_connection = state.get_database_connection().await?;

    let name_stats = database::ssa::get_name_stats(database_connection, query).await?;

    Ok(Json(name_stats))
}
